crates/sha1/src/sha1.rs

```diff
@@ -0,0 +1,406 @@
+where
+    Scalar: PrimeField,
+    CS: ConstraintSystem<Scalar>,
+{
+    assert!(input.len() % 8 == 0);
+
+    let mut padded = input.to_vec();
+    let plen = padded.len() as u64;
+    // append a single '1' bit
+    padded.push(Boolean::constant(true));
+    // append K '0' bits, where K is the minimum number >= 0 such that L + 1 + K + 64 is a multiple of 512
+    while (padded.len() + 64) % 512 != 0 {
+        padded.push(Boolean::constant(false));
+    }
+    // append L as a 64-bit big-endian integer, making the total post-processed length a multiple of 512 bits
+    for b in (0..64).rev().map(|i| (plen >> i) & 1 == 1) {
+        padded.push(Boolean::constant(b));
+    }
+    assert!(padded.len() % 512 == 0);
+
+    let mut cur = get_sha1_iv();
+    for (i, block) in padded.chunks(512).enumerate() {
+        cur = sha1_compression_function(cs.namespace(|| format!("block {}", i)), block, &cur)?;
+    }
+
+    Ok(cur.into_iter().flat_map(|e| e.into_bits_be()).collect())
+}
+
+fn get_sha1_iv() -> Vec<UInt32> {
+    IV.iter().map(|&v| UInt32::constant(v)).collect()
+}
+
+pub fn sha1_compression_function<Scalar, CS>(
+    cs: CS,
+    input: &[Boolean],
+    current_hash_value: &[UInt32],
+) -> Result<Vec<UInt32>, SynthesisError>
+where
+    Scalar: PrimeField,
+    CS: ConstraintSystem<Scalar>,
+{
+    assert_eq!(input.len(), 512);
+    assert_eq!(current_hash_value.len(), 5);
+
+    let mut w = input
+        .chunks(32)
+        .map(UInt32::from_bits_be)
+        .collect::<Vec<_>>();
+
+    // We can save some constraints by combining some of
+    // the constraints in different u32 additions
+    let mut cs = MultiEq::new(cs);
+
+    for i in 16..80 {
+        let cs = &mut cs.namespace(|| format!("w extension {}", i));
+
+        // w[i] := (w[i-3] xor w[i-8] xor w[i-14] xor w[i-16]) leftrotate 1
+        let mut wi = w[i - 3].xor(cs.namespace(|| format!("first xor for w[{i}]")), &w[i - 8])?;
+        wi = wi.xor(
+            cs.namespace(|| format!("second xor for w[{i}]")),
+            &w[i - 14],
+        )?;
+        wi = wi.xor(cs.namespace(|| format!("third xor for w[{i}]")), &w[i - 16])?;
+
+        // rotr(31) is equivalent to leftrotate 1
+        wi = wi.rotr(31);
+        w.push(wi);
+    }
+
+    assert_eq!(w.len(), 80);
+
+    let mut a = current_hash_value[0].clone();
+    let mut b = current_hash_value[1].clone();
+    let mut c = current_hash_value[2].clone();
+    let mut d = current_hash_value[3].clone();
+    let mut e = current_hash_value[4].clone();
+
+    for i in 0..80 {
+        let cs = &mut cs.namespace(|| format!("compression round {}", i));
+
+        let f = if i < 20 {
+            // f = (b and c) or ((not b) and d)
+            UInt32::sha256_ch(cs.namespace(|| "ch"), &b, &c, &d)?
+        } else if !(40..60).contains(&i) {
+            // b xor c xor d
+            b.xor(cs.namespace(|| "b xor c"), &c)?
+                .xor(cs.namespace(|| "b xor c xor d"), &d)?
+        } else {
```

Comment:
Suggestion for a comment saying this branch is for rounds `40 <= t <= 59`.

Reply:
I will add a comment here as well